### PR TITLE
variable: refactor session/global validation (part 3)

### PR DESCRIPTION
**Pull request by morgo**
### What problem does this PR solve?

Problem Summary:

This is a followup to https://github.com/pingcap/tidb/pull/23383 and https://github.com/pingcap/tidb/pull/23566

### What is changed and how it works?

This has a few improvements:

* I noticed that the enum for `tidb_enable_clustered_index` made some changes to checking deprecated functions (I plan to move this functionality into the sysvar struct).
* The validation for `tidb_enable_clustered_index` is also now done in a standard way: where the ordinal values can be used for the values `ON` or `OFF`, and the case will be normalized automatically. There is no need to add values "0" and "1" to the enum values.
* The helper functions that are not related to the sysVar struct are moved from `sysvar.go` to `varsutil.go`.
* ~A new SetGlobal function is introduced (to handle the case that `tidb_enable_clustered_index` needs to generate a warning).~ **Update:** I have moved the warning to the Validation function, and removed the SetGlobal function. It is not required by anything else yet.
* The validation function is moved directly into the sysvar struct.

### Check List 

Tests 

- Unit test (refactor; covered by existing tests).

Side effects

- Harder cherry picks
- Slightly more verbose code in some cases, but more self contained.

### Release note 

- No release note

### CODE REVIEW

File: sessionctx/variable/sysvar.go [hunk 1]
```diff
@@ -349,14 +373,6 @@ func (sv *SysVar) checkInt64SystemVarWithError(value string) (string, error) {
...
 }
 
-// ValidateFromHook calls the anonymous function on the sysvar if it exists.
-func (sv *SysVar) ValidateFromHook(vars *SessionVars, normalizedValue string, originalValue string, scope ScopeFlag) (string, error) {
-	if sv.Validation != nil {
-		return sv.Validation(vars, normalizedValue, originalValue, scope)
-	}
-	return normalizedValue, nil
-}
-
```
Comment (morgo) [inline sessionctx/variable/sysvar.go:359]: This function becomes obsolete because `Validate()` was introduced, replacing varsutil.go: ValidateSetSystemVar.

File: sessionctx/variable/sysvar.go [hunk 2]
```diff
@@ -434,37 +450,29 @@ func init() {
...
-func checkCharacterValid(normalizedValue string, argName string) (string, error) {
-	if normalizedValue == "" {
-		return normalizedValue, errors.Trace(ErrWrongValueForVar.GenWithStackByArgs(argName, "NULL"))
-	}
-	cht, _, err := charset.GetCharsetInfo(normalizedValue)
-	if err != nil {
-		return normalizedValue, errors.Trace(err)
-	}
-	return cht, nil
-}
```
Comment (morgo) [inline sessionctx/variable/sysvar.go:461]: These functions are not related to the struct so they were just moved to varsutil.go.

File: session/session.go [hunk 3]
```diff
@@ -963,29 +963,30 @@ func (s *session) GetGlobalSysVar(name string) (string, error) {
...
+	}
 	if name == variable.TiDBSlowLogMasking {
 		name = variable.TiDBRedactLog
 	}
-	if name == variable.SQLModeVar {
-		value = mysql.FormatSQLModeStr(value)
-		if _, err := mysql.GetSQLMode(value); err != nil {
-			return err
-		}
-	}
```
Comment (morgo) [inline session/session.go:974]: This was moved into validation function on the sysVar struct.

File: sessionctx/variable/session.go [hunk 4]
```diff
@@ -1655,8 +1645,6 @@ func (s *SessionVars) SetSystemVar(name string, val string) error {
 		s.SelectLimit = result
 	case TiDBAllowAutoRandExplicitInsert:
 		s.AllowAutoRandExplicitInsert = TiDBOptOn(val)
-	case TiDBEnableClusteredIndex:
-		s.EnableClusteredIndex = TiDBOptEnableClustered(val)
```
Comment (morgo) [inline sessionctx/variable/session.go:1659]: These were moved into the setsession function on the sysvar struct.

File: sessionctx/variable/sysvar.go [hunk 5]
```diff
@@ -128,8 +129,31 @@ func (sv *SysVar) SetSessionFromHook(s *SessionVars, val string) error {
...
+	normalizedValue, err := sv.validateFromType(vars, value, scope)
+	if err != nil {
+		return normalizedValue, err
+	}
+	// If type validation was successful, call the (optional) validation function
+	if sv.Validation != nil {
+		return sv.Validation(vars, normalizedValue, value, scope)
+	}
+	return normalizedValue, nil
+}
```
Comment (morgo) [inline sessionctx/variable/sysvar.go]: This replaces varsutil.go:ValidateSetSystemVar().

File: sessionctx/variable/varsutil.go [hunk 6]
```diff
@@ -295,29 +323,9 @@ func CheckDeprecationSetSystemVar(s *SessionVars, name string, val string) {
...
-	// Normalize the value and apply validation based on type.
-	// i.e. TypeBool converts 1/on/ON to ON.
-	normalizedValue, err := sv.ValidateFromType(vars, value, scope)
-	if err != nil {
-		return normalizedValue, err
-	}
-	// If type validation was successful, call the (optional) validation function
-	return sv.ValidateFromHook(vars, normalizedValue, value, scope)
-}
-
```
Comment (morgo) [inline sessionctx/variable/varsutil.go:320]: This is replaced by the Validate() function on the sysvar struct.

**Review by morgo — COMMENTED** [review]
I've added some comments to help reviewers.

File: session/session.go [hunk 7]
```diff
@@ -963,29 +963,30 @@ func (s *session) GetGlobalSysVar(name string) (string, error) {
...
+	}
 	name = strings.ToLower(name)
 	// update mysql.tidb if required.
 	if s.varFromTiDBTable(name) {
 		if err = s.setTiDBTableValue(name, sVal); err != nil {
 			return err
 		}
 	}
-	variable.CheckDeprecationSetSystemVar(s.sessionVars, name, sVal)
+	variable.CheckDeprecationSetSystemVar(s.sessionVars, name)
```
Comment (morgo) [inline session/session.go:989]: I will move this into the sysvar struct in a future PR. It should be possible to declaratively state that a sysVar is deprecated.

File: sessionctx/variable/varsutil.go [hunk 8]
```diff
@@ -357,10 +365,10 @@ const (
 
 // TiDBOptEnableClustered converts enable clustered options to ClusteredIndexDefMode.
 func TiDBOptEnableClustered(opt string) ClusteredIndexDefMode {
-	switch {
-	case strings.EqualFold(opt, "ON") || opt == "1":
+	switch opt {
+	case On:
 		return ClusteredIndexDefModeOn
-	case strings.EqualFold(opt, "OFF") || opt == "0":
+	case Off:
```
Comment (morgo) [inline sessionctx/variable/varsutil.go:371]: `strings.EqualFold` is not required because validation also performs normalization.

**Review by xhebox — COMMENTED** [review]
Rest LGTM

**Review by xhebox — APPROVED** [review]
LGTM
